collections/_tutorials/2022-06-20-tutorial_transformer.md

```diff
@@ -0,0 +1,124 @@
+
+$$z_l = MLP(LN(z^{'}_{l})) + z^{'}_{l} \quad \quad \quad \quad l=1 \cdots L$$
+
+&nbsp;
+
+### Norm step
+
+Normalization was inserted into the encoding layer to control the dynamics of the token values before each of the two key steps. The corresponding diagram is given below:
+
+![](/collections/images/transformers/vit_norm_step.jpg)
+
+&nbsp;
+
+### MLP step
+
+The diagram of the MLP procedure is given below:
+
+![](/collections/images/transformers/vit_mlp_step.jpg)
+
+&nbsp;
+
+### Multi-Head Attention block
+
+The Multi-Head Attention (MHA) block is the key element of the encoding layer. It is based on the "qkv" paradigm, but what does "qkv" mean ? 
+Before answering this question, let's zoom in on the MHA block and get an overview of its structure.
+
+![](/collections/images/transformers/vit_mha_overview.jpg)
+
+From this figure, we can see that the key element is the Self-Attention module which outputs $$k$$ head matrices of size $$\mathbb{R}^{(N+1) \times D_h}$$, where $$D_h$$ is usually computed as $$D_h=D/k$$. These head matrices contain useful information computed from attention mechanisms. The second part of the MHA block uses a linear projection to optimally merge the different heads and to output a new sequence of tokens of the same size as the input matrix, $$i.e.$$ $$\mathbb{R}^{(N+1) \times D}$$. This operation is modeled as:
+
+$$ MSA(z) = \left[ {SA}_1(z), \, {SA}_2(z), \cdots, \, {SA}_k(z) \right] \cdot \mathbf{U}_{msa} \quad \quad \quad \quad \mathbf{U}_{msa} \in \mathbb{R}^{(k D_h) \times D}$$
+
+where $${SA}_i(z)$$ represents the output of Head $$i$$ ($$SA$$ stands for Self Attention). Now it is time to investigate the Self-Attention module !
+
+&nbsp;
+
+### Self-Attention module
+
+The Self-Attention module is the core element of the MHA block. The figure below provides an overview of such a module.
+
+![](/collections/images/transformers/vit_self_attention_module.jpg)
+
+&nbsp;
+
+It involves the generation of 3 matrices $$\mathbf{Q}$$, $$\mathbf{K}$$, $$\mathbf{V}$$ of size $$\mathbb{R}^{(N+1) \times D_h}$$ computed from three different linear projection matrices $$\mathbf{U}_{qkv} \in \mathbb{R}^{D \times D_h}$$. This step is modeled as follows:
+
+$$ [\mathbf{Q}, \mathbf{K}, \mathbf{V}] = z \mathbf{U}_{qkv} \quad \quad \quad \quad \mathbf{U}_{qkv} \in \mathbb{R}^{D \times D_h}$$
+
+&nbsp;
+
+$$\mathbf{Q}$$ and $$\mathbf{K}$$ are used to create a self-attention matrix $$\mathbf{A} \in \mathbb{R}^{(N+1) \times (N+1)}$$. Since this matrix is computed from scalar products, it expresses the proximity between token $$i$$ and all token $$j$$. The Softmax operation is applied to each row of $$\mathbf{A}$$ individually to ensure that the sum of the elements of each row is equal to $$1$$. The corresponding operation is:
+
+$$ \mathbf{A} = softmax\left( \frac{\mathbf{Q} \cdot \boldsymbol{K}^T}{\sqrt{D_h}} \right)$$
```

Comment:
Use \mathbf here instead of \boldsymbol for consistency?
```suggestion
$$ \mathbf{A} = softmax\left( \frac{\mathbf{Q} \cdot \mathbf{K}^T}{\sqrt{D_h}} \right)$$
```